File: python/day1/main.py
def main(fn: str = 'input.txt') -> None:
    with open(fn, 'r') as file:
        zeroes = 0
        dial = 50
        size = 100
        for line in file:
            direction = line[0]
            amount = int(line[1:].strip())
            match direction:
                case 'R':
                    dial = (dial + amount) % size
                case 'L':
                    dial = (dial + size - (amount % size)) % size
            if dial == 0:
                zeroes += 1
        print("Part 1:", zeroes)

    with open(fn, 'r') as file:
        dial = 50
        zeroes = 0
        size = 100
        for line in file:
            direction = line[0]
            amount = int(line[1:].strip())
            match direction:
                case 'R':
                    for _ in range(amount):
                        dial = (dial + 1) % size
                        if dial == 0:
                            zeroes += 1
                case 'L':
                    for _ in range(amount):
                        dial = (dial + size - 1) % size
                        if dial == 0:
                            zeroes += 1
        print("Part 2:", zeroes)

if __name__ == '__main__':
    main()

File: go/day1/main.go
package day1

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func Main() {
	input, err := os.ReadFile("/home/luke/dev/advent/2025/inputs/day1.txt")
	if err != nil {
		panic(err)
	}

	// the dial starts at 50, and goes from 0 to 99. We want to count how many times it passes 0.
	dial := uint32(50)
	passes := uint32(0)

	for _, line := range strings.Fields(string(input)) {
		// each line is something like "R15", "L5", etc.
		// get the direction
		if len(line) < 2 {
			continue
		}

		direction := line[0]
		n, err := strconv.ParseUint(line[1:], 10, 32)
		if err != nil {
			panic(err)
		}
		distance := uint32(n)

		// calculate the number of times the dial passes 0
		switch direction {
		case 'R':
			dial += distance
			if dial >= 100 {
				passes += dial / 100
				dial = dial % 100
			}
		case 'L':
			if distance > dial {
				passes += (distance - dial + 99) / 100
			}
			dial = (dial + 100 - (distance % 100)) % 100
		default:
			panic(fmt.Sprintf("invalid direction: %q", direction))
		}
	}

	fmt.Printf("The dial passed 0 %d times\n", passes)
}
